apollo-federation/src/operation/mod.rs

```diff
@@ -2596,45 +2629,46 @@ impl SelectionSet {
                 // schema.
                 let element = ele.rebase_on_or_error(&self.type_position, &self.schema)?;
-                let selection_set = selection_set
-                    .map(|selection_set| {
-                        selection_set.rebase_on(
-                            &element.sub_selection_type_position()?.ok_or_else(|| {
-                                FederationError::internal("unexpected: Element has a selection set with non-composite base type")
-                            })?,
-                            &NamedFragments::default(),
-                            &self.schema,
-                            RebaseErrorHandlingOption::ThrowError,
-                        )
-                    })
-                    .transpose()?;
-                let selection = Selection::from_element(element, selection_set)?;
-                // TODO move the rebasing to add_selection/merge_into
-                if let Some(rebased_selection) = selection.rebase_on(
-                    &self.type_position,
-                    &NamedFragments::default(),
-                    &self.schema,
-                    RebaseErrorHandlingOption::ThrowError,
-                )? {
-                    self.add_selection(rebased_selection)?
+                if selection_set.is_none() || selection_set.is_some_and(|s| s.is_empty()) {
+                    // This is a somewhat common case when dealing with `@key` "conditions" that we can
+                    // end up with trying to add empty sub selection set on a non-leaf node. There is
+                    // nothing to do here - we know will have a node at specified path but currently
+                    // we don't have any sub selections so there is nothing to merge.
+                    // JS code was doing this check in `makeSelectionSet`
+                    if !ele.is_terminal()? {
+                        return Ok(());
+                    } else {
+                        // add leaf
+                        let selection = Selection::from_element(element, None)?;
+                        self.add_selection(&selection)?
+                    }
+                } else {
+                    let selection_set = selection_set
+                        .map(|selection_set| {
+                            selection_set.rebase_on(
+                                &element.sub_selection_type_position()?.ok_or_else(|| {
+                                    FederationError::internal("unexpected: Element has a selection set with non-composite base type")
+                                })?,
+                                &NamedFragments::default(),
+                                &self.schema,
+                                RebaseErrorHandlingOption::ThrowError,
+                            )
+                        })
+                        .transpose()?.map(|selection_set| selection_set.without_unnecessary_fragments());
```

Comment:
```suggestion
                        .transpose()?
                        .map(|selection_set| selection_set.without_unnecessary_fragments());
```
Not sure why linter doesn't complain 🙈